src/envoy/http/authn/authn_utils.cc

```diff
@@ -0,0 +1,108 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "authn_utils.h"
+#include "common/json/json_loader.h"
+#include "src/envoy/http/jwt_auth/jwt.h"
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+
+bool AuthnUtils::GetJWTPayloadFromHeaders(
+    const HeaderMap& headers, const LowerCaseString& jwt_payload_key,
+    istio::authn::JwtPayload* payload) {
+  const HeaderEntry* entry = headers.get(jwt_payload_key);
+  if (!entry) {
+    ENVOY_LOG(error, "{}: no JwtPayloadKey entry {} in the header",
+              __FUNCTION__, jwt_payload_key.get());
+    return false;
+  }
+  std::string value(entry->value().c_str(), entry->value().size());
+  // JwtAuth::Base64UrlDecode() is different from Base64::decode().
+  std::string payload_str = JwtAuth::Base64UrlDecode(value);
+  // Return an empty string if Base64 decode fails.
+  if (payload_str.empty()) {
+    ENVOY_LOG(error, "Invalid {} header, invalid base64: {}",
+              jwt_payload_key.get(), value);
+    return false;
+  }
+  ::google::protobuf::Map< ::std::string, ::std::string>* claims =
+      payload->mutable_claims();
+  try {
+    auto json_obj = Json::Factory::loadFromString(payload_str);
+    ENVOY_LOG(debug, "{}: json object is {}", __FUNCTION__,
+              json_obj->asJsonString());
+    json_obj->iterate([claims, &payload](const std::string& key,
+                                         const Json::Object& obj) -> bool {
+      ENVOY_LOG(debug, "{}: key is {}", __FUNCTION__, key);
+      try {
+        if (key == "aud") {
+          // "aud" can be either string array or string.
+          // Try as string array, read it as empty array if doesn't exist.
+          std::vector<std::string> aud_vector;
+          try {
+            aud_vector = obj.getStringArray(key, true);
+          } catch (Json::Exception& e) {
+            // Try as string
+            try {
+              auto audience = obj.asString();
+              aud_vector.push_back(audience);
+            } catch (Json::Exception& e) {
+              ENVOY_LOG(error,
+                        "{}: an exception is thrown when get audience as "
+                        "string. The exception is: {}",
+                        __FUNCTION__, e.what());
+            }
+          }
+          for (size_t i = 0; i < aud_vector.size(); i++) {
+            payload->add_audiences(aud_vector[i]);
+          }
+        } else {
+          // will throw execption if value type is not string.
+          // In current implementation, only string objects are extracted into
+          // claims.
+          (*claims)[key] = obj.asString();
```

Comment:
so "aud" don't need to add to "claims" map?  who will use claims field?  will they expect that?

Reply:
How about using asJsonString() to take care all types.

Reply:
Per proxy/src/istio/authn/context.proto, 'aud' is a repeated string in JwtPayload and it is not in the claims map. Per the examples in proxy/src/istio/authn/context.proto, the values are plain strings. If there are requirements to encoded them into json strings, we can modify later.

Reply:
I'm ok with this for now. As I am working on the mixer part, I think we haven't handle multiple aud case properly neither. We can also keep all claims as json string, just in case, but it could be done in the next PR.